src/Microsoft.ML.Data/Evaluators/Metrics/ConfusionMatrix.cs

```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Data
+{
+    public sealed class ConfusionMatrix
+    {
+        public double[] PrecisionSums { get; }
+        public double[] RecallSums { get; }
+        public ReadOnlyMemory<char>[] Labels => PredictedLabelNames.ToArray(); //sefilipi: return as string[]
+
+        public double[][] ConfusionTableCounts { get; }
+
+        internal readonly List<ReadOnlyMemory<char>> PredictedLabelNames;
+        internal readonly bool Sampled;
+        internal readonly bool Binary;
+
+        private readonly IHost _host;
+
+        internal ConfusionMatrix(IHost host, double[] precisionSums, double[] recallSums, double[][] confusionTableCounts,
+            List<ReadOnlyMemory<char>> labelNames, bool sampled, bool binary)
+        {
+            _host = host;
+
+            PrecisionSums = precisionSums;
+            RecallSums = recallSums;
+            ConfusionTableCounts = confusionTableCounts;
+            PredictedLabelNames = labelNames;
+            Sampled = sampled;
+            Binary = binary;
+        }
+
+        public override string ToString() => MetricWriter.GetConfusionTableAsString(this, false);
```

Comment:
>MetricWriter.GetConfusionTableAsString [](start = 45, length = 38)

This is dangerous. Do we consider `ToString` overrides breaking changes?

Reply:
You're right, it is a breaking change. 
I'll create a separate API for it. 


"Changing the returned values for a property, field, return or 'out' value, such as the value returned from ToString"
on 
[https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md)



---
In reply to: [273615670](https://github.com/dotnet/machinelearning/pull/3250#discussion_r273615670) [](ancestors = 273615670)